unit_tests/test_cepch_iscsi_charm.py

```diff
@@ -0,0 +1,321 @@
+
+        self.harness._backend = _TestingOPSModelBackend(
+            self.harness._unit_name)
+        self.harness._model = model.Model(
+            self.harness._unit_name,
+            self.harness._meta,
+            self.harness._backend)
+        self.harness._framework = framework.Framework(
+            ":memory:",
+            self.harness._charm_dir,
+            self.harness._meta,
+            self.harness._model)
+        # END Workaround
+
+    def test_init(self):
+        self.harness.begin()
+        self.assertFalse(self.harness.charm.state.target_created)
+        self.assertFalse(self.harness.charm.state.enable_tls)
+        self.assertEqual(self.harness.charm.state.additional_trusted_ips, [])
+
+    def add_cluster_relation(self):
+        rel_id = self.harness.add_relation('cluster', 'ceph-iscsi')
+        self.harness.add_relation_unit(
+            rel_id,
+            'ceph-iscsi/1',
+            {
+                'ingress-address': '10.0.0.2',
+                'gateway_ready': 'True',
+                'gateway_fqdn': 'ceph-iscsi-1.example'
+            }
+        )
+        return rel_id
+
+    @patch('socket.getfqdn')
+    def test_on_create_target_action(self, _getfqdn):
+        _getfqdn.return_value = 'ceph-iscsi-0.example'
+        self.add_cluster_relation()
+        self.harness.begin()
+        action_event = MagicMock()
+        action_event.params = {
+            'iqn': 'iqn.mock.iscsi-gw:iscsi-igw',
+            'gateway-units': 'ceph-iscsi/0 ceph-iscsi/1',
+            'pool-name': 'iscsi-pool',
+            'image-name': 'disk1',
+            'image-size': '5G',
+            'client-initiatorname': 'client-initiator',
+            'client-username': 'myusername',
+            'client-password': 'mypassword'}
+        self.harness.charm.on_create_target_action(action_event)
+        self.gwc.add_gateway_to_target.assert_has_calls([
+            call(
+                'iqn.mock.iscsi-gw:iscsi-igw',
+                '10.0.0.10',
+                'ceph-iscsi-0.example'),
+            call(
+                'iqn.mock.iscsi-gw:iscsi-igw',
+                '10.0.0.2',
+                'ceph-iscsi-1.example')])
+
+        self.gwc.create_pool.assert_called_once_with(
+            'iscsi-pool',
+            'disk1',
+            '5G')
+        self.gwc.add_client_to_target.assert_called_once_with(
+            'iqn.mock.iscsi-gw:iscsi-igw',
+            'client-initiator')
+        self.gwc.add_client_auth.assert_called_once_with(
+            'iqn.mock.iscsi-gw:iscsi-igw',
+            'client-initiator',
+            'myusername',
+            'mypassword')
+        self.gwc.add_disk_to_client.assert_called_once_with(
+            'iqn.mock.iscsi-gw:iscsi-igw',
+            'client-initiator',
+            'iscsi-pool',
+            'disk1')
+
+    @patch.object(charm.secrets, 'choice')
+    def test_on_has_peers(self, _choice):
+        _choice.return_value = 'r'
+        self.add_cluster_relation()
```

Comment:
I think you could move `add_cluster_relation` to a later point (after `harness.begin()`) and avoid having to `emit` an event manually:

```        self.harness.charm.peers.on.has_peers.emit()```

The harness emits joined and changed events itself thus simulating the native Juju behavior.